docs/users/extending-commands.md

```diff
@@ -0,0 +1,103 @@
+<h1>Extending ddev Commands<h1>
+
+Certain ddev commands provide hooks to run tasks before or after the main command executes. These tasks can be defined in the config.yml for your site, and allow for you to automate setup tasks specific to your site. To define command tasks in your configuration, specify the desired command hook as a subfield to `extend-commands`, then provide a list of tasks to run.
+
+Example:
+
+```
+extend-commands:
+  $command-hook:
+    - task: value
+```
+
+## Supported Command Hooks
+
+- `pre-start`: Hooks into "ddev start". Execute tasks before the site environment starts. **Note:** Only `exec-host` tasks can be run successfully for pre-start. See Supported Tasks below for more info.
+- `post-start`: Hooks into "ddev start". Execute tasks after the site environment has started
+- `pre-import-db`: Hooks into "ddev import-db". Execute tasks before database import
+- `post-import-db`: Hooks into "ddev import-db". Execute tasks after database import
+- `pre-import-files`: Hooks into "ddev import-files". Execute tasks before files are imported
+- `post-import-files`: Hooks into "ddev import-files". Execute tasks after files are imported.
+
+## Supported Tasks
+
+- `exec`: Execute a shell command in the web service container.
+
+Value: string providing the command to run. Commands requiring user interaction are not supported.
+
+Example:
+
+_Use drush to clear the drupal cache after database import_
+
+```
+extend-commands:
+  post-import-db:
+    - exec: "drush cc all"
+```
+
+- `exec-host`: Execute a shell command on your system.
+
+Value: string providing the command to run. Commands requiring user interaction are not supported.
+
+Example:
+
+_Run a git pull before starting the site_
+
+```
+extend-commands:
+  pre-start:
+    - exec: "git pull origin master"
```

Comment:
Composer seems like a perfect example of when this would be appropriate, particularly if we detect a D8 app type. This was the very thing that @nerdstein had found as something that would trip up a D8 core contributor now that the vendor files are not included in the repo and only available when you download a package.